Source/WebCore/dom/InternalObserverFirst.cpp

```diff
@@ -0,0 +1,111 @@
+/*
+* Copyright (C) 2024 Marais Rossouw <me@marais.co>. All rights reserved.
+*
+* Redistribution and use in source and binary forms, with or without
+* modification, are permitted provided that the following conditions
+* are met:
+* 1. Redistributions of source code must retain the above copyright
+*    notice, this list of conditions and the following disclaimer.
+* 2. Redistributions in binary form must reproduce the above copyright
+*    notice, this list of conditions and the following disclaimer in the
+*    documentation and/or other materials provided with the distribution.
+*
+* THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+* AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+* THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+* PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+* BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+* CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+* SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+* INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+* CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+* ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+* THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#include "config.h"
+#include "InternalObserverFirst.h"
+
+#include "AbortController.h"
+#include "AbortSignal.h"
+#include "Exception.h"
+#include "ExceptionCode.h"
+#include "InternalObserver.h"
+#include "JSDOMPromiseDeferred.h"
+#include "Observable.h"
+#include "ScriptExecutionContext.h"
+#include "SubscribeOptions.h"
+#include "Subscriber.h"
+#include "SubscriberCallback.h"
+#include <JavaScriptCore/JSCJSValueInlines.h>
+
+namespace WebCore {
+
+class InternalObserverFirst final : public InternalObserver {
+public:
+    static Ref<InternalObserverFirst> create(ScriptExecutionContext& context, Ref<AbortSignal>&& signal, Ref<DeferredPromise>&& promise)
+    {
+        Ref internalObserver = adoptRef(*new InternalObserverFirst(context, WTFMove(signal), WTFMove(promise)));
+        internalObserver->suspendIfNeeded();
+        return internalObserver;
+    }
+
+private:
+    void next(JSC::JSValue value) final
+    {
+        Ref { m_promise }->resolve<IDLAny>(value);
+        Ref { m_signal }->signalAbort(JSC::jsUndefined());
```

Comment:
I wonder if this is really correct to pass `JSC::jsUndefined()` here, I suspect we may want `toJS(&globalObject, &globalObject, DOMException::create(ExceptionCode::AbortError)` instead. 

The spec says to signal abort controller. However, you opted NOT to use AbortController and use AbortSignal directly. Note how `AbortController::signalAbort()` is implemented:
```
void AbortController::abort(JSDOMGlobalObject& globalObject, JSC::JSValue reason)
{
    ASSERT(reason);
    if (reason.isUndefined())
        reason = toJS(&globalObject, &globalObject, DOMException::create(ExceptionCode::AbortError));

    protectedSignal()->signalAbort(reason);
}
```

Note the fallback for `reason`. This matches step 2 of https://dom.spec.whatwg.org/#abortsignal-signal-abort. 
I suspect our `reason` fallback should be in `AbortSignal::signalAbort()` instead of `AbortController::abort()` to match the spec exactly.

Reply:
> I suspect our reason fallback should be in AbortSignal::signalAbort() instead of AbortController::abort() to match the spec exactly.

I can raise a bug/fix PR for this, and explore what that means aside from this change. For this PR specifically, I will use `AbortController` instead of AbortSignal. Think that way we aren't leaking another spec's detail here as well.

Reply:
@cdumez I have raised a bug issue [here](https://bugs.webkit.org/show_bug.cgi?id=281961) and have started at an implementation fix for it. Ill continue the conversation over on the new diff for that.